src/common/ical.rs

```diff
@@ -0,0 +1,142 @@
+use icalendar::{Calendar, Component};
+
+use crate::handlers::error::ResourceError;
+
+pub async fn parse_ical_uri(
+    user_id: String,
+    ical_uri: String,
+) -> Result<Vec<crate::entity::event::EventEntity>, ResourceError> {
+    // Request source
+    let client = reqwest::Client::new();
+    let ical_body = client
+        .get(&ical_uri.clone())
+        .send()
+        .await
+        .map_err(|e| ResourceError::NetworkError)?
+        .text()
+        .await
+        .map_err(|e| ResourceError::FailedParse("Failed to read source response".to_string()))?;
+
+    let events = parse_ical(user_id, ical_body).await?;
+
+    Ok(events)
+}
+
+async fn parse_ical(
+    user_id: String,
+    ical: String,
+) -> Result<Vec<crate::entity::event::EventEntity>, ResourceError> {
+    let calendar = ical
+        .parse::<icalendar::Calendar>()
+        .map_err(|e| ResourceError::FailedParse("calendar body".to_string()))?;
+
+    let mut result = Vec::new();
+
+    for component in calendar.iter() {
+        let event = match component.as_event() {
+            Some(event) => event,
+            None => continue,
+        };
+
+        let title = event.get_summary().ok_or(ResourceError::FailedParse(
+            "event title is empty".to_string(),
+        ))?;
+
+        let description = match event.get_description() {
+            Some(description) => description,
+            None => "",
+        };
+
+        let start = event
+            .get_start()
+            .ok_or(ResourceError::FailedParse(
+                "event start is empty".to_string(),
+            ))?
+            .to_utc()?;
+
+        let end = event
+            .get_end()
+            .ok_or(ResourceError::FailedParse("event end is empty".to_string()))?
+            .to_utc()?;
+
+        let location = event.property_value("LOCATION").unwrap_or("");
+
+        let id = event
+            .get_uid()
+            .ok_or(ResourceError::FailedParse("event id is empty".to_string()))?;
+
+        let event = crate::entity::event::EventEntity {
+            id: None,
+            user_id: user_id.to_string(),
+            title: title.to_string(),
+            description: description.to_string(),
+            start: start,
+            end: end,
+            all_day: start - end == chrono::Duration::min_value(),
+            location: location.to_string(),
+            original: Some(id.to_string()),
+        };
+
+        result.push(event);
+    }
+
+    Ok(result)
+}
+
+trait ToUtc {
+    fn to_utc(self) -> Result<chrono::DateTime<chrono::Utc>, ResourceError>;
+}
+
+impl ToUtc for icalendar::DatePerhapsTime {
+    fn to_utc(self) -> Result<chrono::DateTime<chrono::Utc>, ResourceError> {
```

Comment:
Might it be possible to use the library [dateparser ](https://github.com/waltzofpearls/dateparser)for this? This supports more options and will be used in the create event feature.

Reply:
Might be. Would have to format the `icalendar::DatePerhapsTime` to a `String` and have that be parsed back to a `chrono::DateTime<chrono::Utc>` using the lib. This function currently has trouble applying the timezone *if* one has been provided by the iCal body. Would need future work either way.